File: Sources/AvivCore/AppCommandCatalog.swift
import AppKit

public struct AppMenuSpec: Equatable {
    public let title: String
    public let entries: [AppMenuEntry]

    public init(title: String, entries: [AppMenuEntry]) {
        self.title = title
        self.entries = entries
    }
}

public enum AppMenuEntry: Equatable {
    case command(AppCommandSpec)
    case separator
    case submenu(title: String, entries: [AppMenuEntry])
}

public struct AppCommandSpec: Equatable {
    public enum TargetRoute: Equatable {
        case application
        case appDelegate
        case firstResponder
    }

    public let identifier: String
    public let title: String
    public let actionName: String
    public let keyEquivalent: String
    public let modifierMask: NSEvent.ModifierFlags
    public let targetRoute: TargetRoute
    public let tag: Int

    public init(
        identifier: String,
        title: String,
        actionName: String,
        keyEquivalent: String = "",
        modifierMask: NSEvent.ModifierFlags = [],
        targetRoute: TargetRoute,
        tag: Int = 0
    ) {
        self.identifier = identifier
        self.title = title
        self.actionName = actionName
        self.keyEquivalent = keyEquivalent
        self.modifierMask = modifierMask
        self.targetRoute = targetRoute
        self.tag = tag
    }
}

public enum AppCommandCatalog {
    public static let menus: [AppMenuSpec] = [
        AppMenuSpec(title: "Aviv", entries: [
            .command(appCommand("about", "About Aviv", "orderFrontStandardAboutPanel:")),
            .separator,
            .command(appCommand("hide", "Hide Aviv", "hide:", key: "h", modifiers: [.command])),
            .command(appCommand("hideOthers", "Hide Others", "hideOtherApplications:", key: "h", modifiers: [.command, .option])),
            .command(appCommand("showAll", "Show All", "unhideAllApplications:")),
            .separator,
            .command(appCommand("quit", "Quit Aviv", "terminate:", key: "q", modifiers: [.command]))
        ]),
        AppMenuSpec(title: "File", entries: [
            .command(delegateCommand("new", "New", "newDocument:", key: "n")),
            .command(delegateCommand("newTab", "New Tab", "newTab:", key: "t")),
            .command(delegateCommand("open", "Open...", "openDocument:", key: "o")),
            .separator,
            .command(delegateCommand("close", "Close", "closeDocument:", key: "w")),
            .separator,
            .command(delegateCommand("save", "Save", "saveDocument:", key: "s")),
            .command(delegateCommand("saveAs", "Save As...", "saveDocumentAs:", key: "S", modifiers: [.command, .shift])),
            .command(delegateCommand("revert", "Revert to Saved", "revertDocumentToSaved:")),
            .separator,
            .command(delegateCommand("pageSetup", "Page Setup...", "pageSetup:", key: "P", modifiers: [.command, .shift])),
            .command(delegateCommand("print", "Print...", "printDocument:", key: "p"))
        ]),
        AppMenuSpec(title: "Edit", entries: [
            .command(responderCommand("undo", "Undo", "undo:", key: "z")),
            .command(responderCommand("redo", "Redo", "redo:", key: "Z", modifiers: [.command, .shift])),
            .separator,
            .command(responderCommand("cut", "Cut", "cut:", key: "x")),
            .command(responderCommand("copy", "Copy", "copy:", key: "c")),
            .command(responderCommand("paste", "Paste", "paste:", key: "v")),
            .command(responderCommand("pasteAndMatchStyle", "Paste and Match Style", "pasteAsPlainText:", key: "V", modifiers: [.command, .option, .shift])),
            .command(responderCommand("delete", "Delete", "delete:")),
            .separator,
            .command(responderCommand("selectAll", "Select All", "selectAll:", key: "a")),
            .separator,
            .submenu(title: "Find", entries: [
                .command(findCommand("find", "Find...", NSTextFinder.Action.showFindInterface, key: "f")),
                .command(findCommand("findAndReplace", "Find and Replace...", NSTextFinder.Action.showReplaceInterface, key: "f", modifiers: [.command, .option])),
                .command(findCommand("findNext", "Find Next", NSTextFinder.Action.nextMatch, key: "g")),
                .command(findCommand("findPrevious", "Find Previous", NSTextFinder.Action.previousMatch, key: "G", modifiers: [.command, .shift])),
                .command(findCommand("useSelectionForFind", "Use Selection for Find", NSTextFinder.Action.setSearchString, key: "e")),
                .command(responderCommand("jumpToSelection", "Jump to Selection", "centerSelectionInVisibleArea:", key: "j"))
            ]),
            .submenu(title: "Spelling and Grammar", entries: [
                .command(responderCommand("showSpelling", "Show Spelling and Grammar", "showGuessPanel:", key: ":", modifiers: [.command])),
                .command(responderCommand("checkSpelling", "Check Document Now", "checkSpelling:", key: ";", modifiers: [.command]))
            ])
        ]),
        AppMenuSpec(title: "View", entries: [
            .command(delegateCommand("actualSize", "Actual Size", "resetTextSize:", key: "0")),
            .command(delegateCommand("zoomIn", "Zoom In", "increaseTextSize:", key: "+")),
            .command(delegateCommand("zoomOut", "Zoom Out", "decreaseTextSize:", key: "-"))
        ]),
        AppMenuSpec(title: "Format", entries: [
            .command(delegateCommand("bold", "Bold", "toggleBold:", key: "b")),
            .command(delegateCommand("italic", "Italic", "toggleItalic:", key: "i")),
            .command(delegateCommand("code", "Code", "toggleCode:", key: "`")),
            .separator,
            .command(delegateCommand("heading1", "Heading 1", "heading1:", key: "1")),
            .command(delegateCommand("heading2", "Heading 2", "heading2:", key: "2"))
        ]),
        AppMenuSpec(title: "Window", entries: [
            .command(responderCommand("minimize", "Minimize", "performMiniaturize:", key: "m")),
            .command(responderCommand("zoomWindow", "Zoom", "performZoom:")),
            .separator,
            .command(responderCommand("showPreviousTab", "Show Previous Tab", "selectPreviousTab:", key: "{", modifiers: [.command, .shift])),
            .command(responderCommand("showNextTab", "Show Next Tab", "selectNextTab:", key: "}", modifiers: [.command, .shift])),
            .command(responderCommand("moveTabToNewWindow", "Move Tab to New Window", "moveTabToNewWindow:")),
            .command(responderCommand("mergeAllWindows", "Merge All Windows", "mergeAllWindows:")),
            .separator,
            .command(appCommand("bringAllToFront", "Bring All to Front", "arrangeInFront:"))
        ])
    ]

    public static var commands: [AppCommandSpec] {
        menus.flatMap { commands(in: $0.entries) }
    }

    public static func command(identifier: String) -> AppCommandSpec? {
        commands.first { $0.identifier == identifier }
    }

    private static func commands(in entries: [AppMenuEntry]) -> [AppCommandSpec] {
        entries.flatMap { entry -> [AppCommandSpec] in
            switch entry {
            case .command(let command):
                return [command]
            case .separator:
                return []
            case .submenu(_, let entries):
                return commands(in: entries)
            }
        }
    }

    private static func appCommand(
        _ identifier: String,
        _ title: String,
        _ actionName: String,
        key: String = "",
        modifiers: NSEvent.ModifierFlags = []
    ) -> AppCommandSpec {
        AppCommandSpec(
            identifier: identifier,
            title: title,
            actionName: actionName,
            keyEquivalent: key,
            modifierMask: modifiers,
            targetRoute: .application
        )
    }

    private static func delegateCommand(
        _ identifier: String,
        _ title: String,
        _ actionName: String,
        key: String = "",
        modifiers: NSEvent.ModifierFlags = [.command]
    ) -> AppCommandSpec {
        AppCommandSpec(
            identifier: identifier,
            title: title,
            actionName: actionName,
            keyEquivalent: key,
            modifierMask: key.isEmpty ? [] : modifiers,
            targetRoute: .appDelegate
        )
    }

    private static func responderCommand(
        _ identifier: String,
        _ title: String,
        _ actionName: String,
        key: String = "",
        modifiers: NSEvent.ModifierFlags = [.command]
    ) -> AppCommandSpec {
        AppCommandSpec(
            identifier: identifier,
            title: title,
            actionName: actionName,
            keyEquivalent: key,
            modifierMask: key.isEmpty ? [] : modifiers,
            targetRoute: .firstResponder
        )
    }

    private static func findCommand(
        _ identifier: String,
        _ title: String,
        _ action: NSTextFinder.Action,
        key: String,
        modifiers: NSEvent.ModifierFlags = [.command]
    ) -> AppCommandSpec {
        AppCommandSpec(
            identifier: identifier,
            title: title,
            actionName: "performFindPanelAction:",
            keyEquivalent: key,
            modifierMask: modifiers,
            targetRoute: .firstResponder,
            tag: action.rawValue
        )
    }
}

public enum AppMenuBuilder {
    public static func makeMenu(
        from menus: [AppMenuSpec] = AppCommandCatalog.menus,
        targetResolver: (AppCommandSpec.TargetRoute) -> AnyObject?
    ) -> NSMenu {
        let mainMenu = NSMenu()
        for menuSpec in menus {
            let menuItem = NSMenuItem()
            let menu = NSMenu(title: menuSpec.title)
            menuItem.submenu = menu
            mainMenu.addItem(menuItem)
            append(menuSpec.entries, to: menu, targetResolver: targetResolver)
        }
        return mainMenu
    }

    private static func append(
        _ entries: [AppMenuEntry],
        to menu: NSMenu,
        targetResolver: (AppCommandSpec.TargetRoute) -> AnyObject?
    ) {
        for entry in entries {
            switch entry {
            case .separator:
                menu.addItem(.separator())
            case .command(let command):
                menu.addItem(menuItem(for: command, targetResolver: targetResolver))
            case .submenu(let title, let entries):
                let item = NSMenuItem()
                item.title = title
                let submenu = NSMenu(title: title)
                item.submenu = submenu
                menu.addItem(item)
                append(entries, to: submenu, targetResolver: targetResolver)
            }
        }
    }

    private static func menuItem(
        for command: AppCommandSpec,
        targetResolver: (AppCommandSpec.TargetRoute) -> AnyObject?
    ) -> NSMenuItem {
        let item = NSMenuItem(
            title: command.title,
            action: Selector(command.actionName),
            keyEquivalent: command.keyEquivalent
        )
        item.identifier = NSUserInterfaceItemIdentifier(command.identifier)
        item.keyEquivalentModifierMask = command.modifierMask
        item.target = targetResolver(command.targetRoute)
        item.tag = command.tag
        return item
    }
}

File: Windows/src/Aviv.Windows.Core/AppCommandCatalog.cs
namespace Aviv.Windows.Core;

[Flags]
public enum AvivKeyModifiers
{
    None = 0,
    Ctrl = 1,
    Shift = 2,
    Alt = 4
}

public enum AppCommandRoute
{
    Application,
    AppViewModel,
    Editor
}

public sealed record AppCommandSpec(
    string Identifier,
    string Title,
    string ActionName,
    string Key = "",
    AvivKeyModifiers Modifiers = AvivKeyModifiers.None,
    AppCommandRoute Route = AppCommandRoute.Editor,
    int Tag = 0);

public abstract record AppMenuEntry
{
    public sealed record Command(AppCommandSpec Spec) : AppMenuEntry;
    public sealed record Separator : AppMenuEntry;
    public sealed record Submenu(string Title, IReadOnlyList<AppMenuEntry> Entries) : AppMenuEntry;
}

public sealed record AppMenuSpec(string Title, IReadOnlyList<AppMenuEntry> Entries);

public static class AppCommandCatalog
{
    public static IReadOnlyList<AppMenuSpec> Menus { get; } =
    [
        new("Aviv",
        [
            App("about", "About Aviv", "ShowAbout"),
            Sep(),
            App("quit", "Quit Aviv", "Quit", "Q", AvivKeyModifiers.Ctrl)
        ]),
        new("File",
        [
            ViewModel("new", "New", "NewDocument", "N"),
            ViewModel("newTab", "New Tab", "NewTab", "T"),
            ViewModel("open", "Open...", "OpenDocument", "O"),
            Sep(),
            ViewModel("close", "Close", "CloseDocument", "W"),
            Sep(),
            ViewModel("save", "Save", "SaveDocument", "S"),
            ViewModel("saveAs", "Save As...", "SaveDocumentAs", "S", AvivKeyModifiers.Ctrl | AvivKeyModifiers.Shift),
            ViewModel("revert", "Revert to Saved", "RevertDocument"),
            Sep(),
            ViewModel("pageSetup", "Page Setup...", "PageSetup", "P", AvivKeyModifiers.Ctrl | AvivKeyModifiers.Shift),
            ViewModel("print", "Print...", "PrintDocument", "P")
        ]),
        new("Edit",
        [
            Editor("undo", "Undo", "Undo", "Z"),
            Editor("redo", "Redo", "Redo", "Z", AvivKeyModifiers.Ctrl | AvivKeyModifiers.Shift),
            Sep(),
            Editor("cut", "Cut", "Cut", "X"),
            Editor("copy", "Copy", "Copy", "C"),
            Editor("paste", "Paste", "Paste", "V"),
            Editor("pasteAndMatchStyle", "Paste and Match Style", "PastePlainText", "V", AvivKeyModifiers.Ctrl | AvivKeyModifiers.Alt | AvivKeyModifiers.Shift),
            Editor("delete", "Delete", "Delete"),
            Sep(),
            Editor("selectAll", "Select All", "SelectAll", "A"),
            Sep(),
            Submenu("Find",
            [
                Editor("find", "Find...", "ShowFind", "F"),
                Editor("findAndReplace", "Find and Replace...", "ShowReplace", "F", AvivKeyModifiers.Ctrl | AvivKeyModifiers.Alt),
                Editor("findNext", "Find Next", "FindNext", "G"),
                Editor("findPrevious", "Find Previous", "FindPrevious", "G", AvivKeyModifiers.Ctrl | AvivKeyModifiers.Shift),
                Editor("useSelectionForFind", "Use Selection for Find", "UseSelectionForFind", "E"),
                Editor("jumpToSelection", "Jump to Selection", "JumpToSelection", "J")
            ])
        ]),
        new("View",
        [
            ViewModel("actualSize", "Actual Size", "ResetTextSize", "0"),
            ViewModel("zoomIn", "Zoom In", "IncreaseTextSize", "+"),
            ViewModel("zoomOut", "Zoom Out", "DecreaseTextSize", "-")
        ]),
        new("Format",
        [
            ViewModel("bold", "Bold", "ToggleBold", "B"),
            ViewModel("italic", "Italic", "ToggleItalic", "I"),
            ViewModel("code", "Code", "ToggleCode", "`"),
            Sep(),
            ViewModel("heading1", "Heading 1", "Heading1", "1"),
            ViewModel("heading2", "Heading 2", "Heading2", "2")
        ]),
        new("Window",
        [
            ViewModel("minimize", "Minimize", "MinimizeWindow", "M"),
            ViewModel("zoomWindow", "Zoom", "ZoomWindow"),
            Sep(),
            ViewModel("showPreviousTab", "Show Previous Tab", "ShowPreviousTab", "Tab", AvivKeyModifiers.Ctrl | AvivKeyModifiers.Shift),
            ViewModel("showNextTab", "Show Next Tab", "ShowNextTab", "Tab"),
            ViewModel("moveTabToNewWindow", "Move Tab to New Window", "MoveTabToNewWindow"),
            ViewModel("mergeAllWindows", "Merge All Windows", "MergeAllWindows"),
            Sep(),
            App("bringAllToFront", "Bring All to Front", "BringAllToFront")
        ])
    ];

    public static IReadOnlyList<AppCommandSpec> Commands => Menus.SelectMany(menu => CommandsIn(menu.Entries)).ToArray();

    public static AppCommandSpec? Command(string identifier)
    {
        return Commands.FirstOrDefault(command => command.Identifier == identifier);
    }

    private static IEnumerable<AppCommandSpec> CommandsIn(IEnumerable<AppMenuEntry> entries)
    {
        foreach (var entry in entries)
        {
            switch (entry)
            {
                case AppMenuEntry.Command command:
                    yield return command.Spec;
                    break;
                case AppMenuEntry.Submenu submenu:
                    foreach (var child in CommandsIn(submenu.Entries))
                    {
                        yield return child;
                    }
                    break;
            }
        }
    }

    private static AppMenuEntry.Command App(string identifier, string title, string action, string key = "", AvivKeyModifiers modifiers = AvivKeyModifiers.None)
    {
        return new(new AppCommandSpec(identifier, title, action, key, modifiers, AppCommandRoute.Application));
    }

    private static AppMenuEntry.Command ViewModel(string identifier, string title, string action, string key = "", AvivKeyModifiers modifiers = AvivKeyModifiers.Ctrl)
    {
        return new(new AppCommandSpec(identifier, title, action, key, key.Length == 0 ? AvivKeyModifiers.None : modifiers, AppCommandRoute.AppViewModel));
    }

    private static AppMenuEntry.Command Editor(string identifier, string title, string action, string key = "", AvivKeyModifiers modifiers = AvivKeyModifiers.Ctrl)
    {
        return new(new AppCommandSpec(identifier, title, action, key, key.Length == 0 ? AvivKeyModifiers.None : modifiers, AppCommandRoute.Editor));
    }

    private static AppMenuEntry.Separator Sep() => new();

    private static AppMenuEntry.Submenu Submenu(string title, IReadOnlyList<AppMenuEntry> entries) => new(title, entries);
}
